cmd/kubeadm/app/phases/selfhosting/selfhosting.go

```diff
@@ -0,0 +1,143 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package selfhosting
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kuberuntime "k8s.io/apimachinery/pkg/runtime"
+	clientset "k8s.io/client-go/kubernetes"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	"k8s.io/kubernetes/pkg/api"
+)
+
+const (
+	kubeAPIServer         = "kube-apiserver"
+	kubeControllerManager = "kube-controller-manager"
+	kubeScheduler         = "kube-scheduler"
+
+	selfHostingLabelValuePrefix = "self-hosted-"
+)
+
+// CreateSelfHostedControlPlane is responsible for turning a Static Pod-hosted control plane to a self-hosted one
+// It achieves that task this way:
+// 1. Load the Static Pod specification from disk (from /etc/kubernetes/manifests)
+// 2. Extract the PodSpec from that Static Pod specification
+// 3. Mutate the PodSpec to be compatible with self-hosting (add the right labels, taints, etc. so it can schedule correctly)
+// 4. Build a new DaemonSet object for the self-hosted component in question. Use the PodSpec above mentioned PodSpec
+// 5. Create the DaemonSet resource. Wait for the Pod to become Running.
+// 6. Remove the Static Pod manifest file. The kubelet will stop the original Static Pod-hosted component that was running.
+// 7. The self-hosted container should now step up and take over.
+// 8. In order to avoid race conditions, we're still making sure the API /healthz endpoint is healthy
+// 9. Do that for the kube-apiserver, kube-controller-manager and kube-scheduler in a loop
+func CreateSelfHostedControlPlane(client *clientset.Clientset) error {
+	components := []string{kubeAPIServer, kubeControllerManager, kubeScheduler}
+
+	for _, componentName := range components {
+		start := time.Now()
+		manifestPath := buildStaticManifestFilepath(componentName)
+
+		// Load the Static Pod file in order to be able to create a self-hosted variant of that file
+		podSpec, err := loadPodSpecFromFile(manifestPath)
+		if err != nil {
+			return err
+		}
+
+		// Build a DaemonSet object from the loaded PodSpec
+		ds := buildDaemonSet(componentName, podSpec)
+
+		// Create the DaemonSet in the API Server
+		if _, err := client.ExtensionsV1beta1().DaemonSets(metav1.NamespaceSystem).Create(ds); err != nil {
+			return fmt.Errorf("failed to create self-hosted %q daemonset [%v]", componentName, err)
+		}
+
+		// Wait for the self-hosted component to come up
+		kubeadmutil.WaitForPodsWithLabel(client, componentName, buildSelfHostedWorkloadLabelQuery(componentName))
+
+		// Remove the old Static Pod manifest
```

Comment:
Discussed with @jamiehannaford on Slack and decided to hold off that for now, and re-order the sequence instead. This is fine as-is now and we'll change it if we change it in the future.